tlspuffin/src/tls/rustls/msgs/message.rs

```diff
@@ -64,6 +64,43 @@ impl MessagePayload {
         .ok_or_else(|| Error::corrupt_message(typ))*/
     }
 
+    /// Extract multiple messages payloads from one ApplicationData message
+    pub fn multiple_new(
+        typ: ContentType,
+        vers: ProtocolVersion,
+        payload: Payload,
+    ) -> Result<Vec<Self>, Error> {
+        let fallback_payload = payload.clone();
+        let mut r = Reader::init(&payload.0);
+        let mut parsed: Vec<Self> = vec![];
+        while r.any_left() {
```

Comment:
This is pretty crazy, did not know this. Can we maybe turn that behavior off in boringssl? If not should we maybe rework more of the algebra here to model the real TLS here that can include multiple messages per ApplicationData?

Reply:
This line here indicates that plaintext only contains a single message: 

https://datatracker.ietf.org/doc/html/rfc8446#section-5.2:~:text=content%3A%20%20The%20TLSPlaintext.fragment%20value%2C%20containing%20the%20byte%0A%20%20%20%20%20%20encoding%20of%20a%20handshake%20or%20an%20alert%20message%2C%20or%20the%20raw%20bytes%20of%0A%20%20%20%20%20%20the%20application%27s%20data%20to%20send.

Reply:
We likely can switch to NULL cipher in boringssl: https://boringssl.googlesource.com/boringssl/+/refs/heads/master/ssl/s3_both.cc#189

null ciphers should be available with boringssl: https://github.com/google/boringssl/blob/master/FUZZING.md#fuzzer-mode

Reply:
> This line here indicates that plaintext only contains a single message:
> 
> https://datatracker.ietf.org/doc/html/rfc8446#section-5.2:~:text=content%3A%20%20The%20TLSPlaintext.fragment%20value%2C%20containing%20the%20byte%0A%20%20%20%20%20%20encoding%20of%20a%20handshake%20or%20an%20alert%20message%2C%20or%20the%20raw%20bytes%20of%0A%20%20%20%20%20%20the%20application%27s%20data%20to%20send.

From section 5.1 : 
```
 Handshake messages MAY be coalesced into a single TLSPlaintext record
   or fragmented across several records, provided that:

   -  Handshake messages MUST NOT be interleaved with other record
      types.  That is, if a handshake message is split over two or more
      records, there MUST NOT be any other records between them.

   -  Handshake messages MUST NOT span key changes.  Implementations
      MUST verify that all messages immediately preceding a key change
      align with a record boundary; if not, then they MUST terminate the
      connection with an "unexpected_message" alert.  Because the
      ClientHello, EndOfEarlyData, ServerHello, Finished, and KeyUpdate
      messages can immediately precede a key change, implementations
      MUST send these messages in alignment with a record boundary.
```

Reply:
Sorry, I was not very clear. The code here suggests that using the NULL cipher avoid putting messages into a single record: https://boringssl.googlesource.com/boringssl/+/refs/heads/master/ssl/s3_both.cc#189

(a small patch would likely also make it work without null ciphers)